runtime/sak.c

```diff
@@ -70,29 +170,40 @@ static void encode_C_literal(const char_os * path)
       printf("\\n");
     } else if (c == '\r') {
       printf("\\r");
-#ifndef _WIN32
-    /* On Unix, nothing else needs escaping */
     } else {
-      putchar(c);
+#ifndef _WIN32
+      if (utf8) {
+        /* On Unix, nothing else needs escaping */
+        putchar(c);
+      } else {
+        path += print_utf8_as_utf16(path) - 1;
+      }
 #else
-    /* On Windows, allow 7-bit printable characters to be displayed literally
-       and escape everything else (using the older \x notation for increased
-       compatibility, rather than the newer \U. */
-    } else if (c < 0x80 && iswprint(c)) {
-      putwchar(c);
-    } else {
-      printf("\\x%04x", c);
+      if (utf8) {
+        path += print_utf16_as_utf8(path) - 1;
+      } else {
+        /* On Windows, allow 7-bit printable characters to be displayed
+           literally and escape everything else. */
+        if (printable(c)) {
+          putwchar(c);
+        } else {
+          printf(F16, c);
+        }
+      }
 #endif
     }
+    path++;
   }
 
   putchar('"');
 }
 
 int main_os(int argc, char_os **argv)
 {
-  if (argc == 3 && !strcmp_os(argv[1], T("encode-C-literal"))) {
-    encode_C_literal(argv[2]);
+  if (argc == 3 && !strcmp_os(argv[1], T("encode-C-utf8-literal"))) {
+    encode_C_literal(argv[2], 1);
+  } else if (argc == 3 && !strcmp_os(argv[1], T("encode-C-utf16-literal"))) {
+    encode_C_literal(argv[2], 0);
```

Comment:
```suggestion
  if (argc == 3 && 0 == strcmp_os(argv[1], T("encode-C-utf8-literal"))) {
    encode_C_literal(argv[2], true);
  } else if (argc == 3 && 0 == strcmp_os(argv[1], T("encode-C-utf16-literal"))) {
    encode_C_literal(argv[2], false);
```

Reply:
Regarding `strcmp`, all alternatives, namely `!strcmp`, `0 == strcmp` and `strcmp == 0` appear in the code base. So I’ll keep it as it was in `sak.c` (which happens to have my preference...).